### ISSUE TITLE
ttnn.fused_rms_minimal hangs on Blackhole p300c (2x2 mesh, 4096-wide rms-norm)

### ISSUE THREAD

**Issue by rpavlovicTT** [tenstorrent/tt-metal#45175]
## Summary

`ttnn.fused_rms_minimal` (in `ttnn/cpp/ttnn/operations/experimental/ccl/rms_allgather/`) **hangs indefinitely on Blackhole p300c**, on a 2×2 mesh, with the exact tensor and program-config shape produced by Llama-3.1-70B TP-4 decode. The non-fused decomposition (pre/post + manual `ttnn.all_gather`) works correctly on the same hardware, so the issue is isolated to this fused kernel.

**Reproduces deterministically** with the standalone script below — no compiler / tt-mlir / tt-xla involvement needed.

## Hardware

- QB2 host, 4× Blackhole p300c chips
- Worker grid per chip: 10×11 (harvested)
- 2×2 mesh, `cluster_axis=0` (gather across the Y-axis pair, 2 devices per all-gather)

## Captured op config

```
ttnn.fused_rms_minimal(input, gamma, stats, device)
  cluster_axis = 0
  epsilon = 9.99999974e-6
  compute_config = { math_fidelity=hifi4, fp32_dest_acc_en=true, packer_l1_acc=true }
  program_config = LayerNormShardedMultiCoreProgramConfig(
      compute_with_storage_grid_size = (11, 6),
      subblock_w = 1, block_h = 1, block_w = 2, inplace = false)

input  : 1x1x32x4096 bf16, width-sharded L1, 64 cores at
         [(0,0)-(10,4), (0,5)-(8,5)]   shard 32x64 = 1x2 tiles
weight : 128x32      bf16, interleaved DRAM, ROW_MAJOR (gamma reshaped from 1D)
stats  : 1x1x32x32   f32,  width-sharded L1, core (0,0)
output : same layout as input
```

## Symptoms

- The call to `ttnn.fused_rms_minimal` never returns.
- `tt-triage` shows the hung op stopped on all 4 devices, hung tensor `[1,1,32,4096] bf16 width-sharded L1`.
- NoC counters show small kernel-internal mismatches (e.g. atomic_resp 960 vs 961, write_ack 228455 vs 228453) — back-pressure inside the fused kernel rather than the read-starvation pattern seen with other ring all-gather fabric stalls.
- `tt-smi -r` recovers cleanly.

## Standalone repro

```python
import torch
import ttnn


def main():
    ttnn.set_fabric_config(ttnn.FabricConfig.FABRIC_1D_RING)
    mesh_device = ttnn.open_mesh_device(ttnn.MeshShape(2, 2))

    try:
        input_core_grid = ttnn.CoreRangeSet([
            ttnn.CoreRange(ttnn.CoreCoord(0, 0), ttnn.CoreCoord(10, 4)),
            ttnn.CoreRange(ttnn.CoreCoord(0, 5), ttnn.CoreCoord(8, 5)),
        ])
        assert input_core_grid.num_cores() == 64

        input_mem_cfg = ttnn.create_sharded_memory_config(
            shape=(32, 64),
            core_grid=input_core_grid,
            strategy=ttnn.ShardStrategy.WIDTH,
            orientation=ttnn.ShardOrientation.ROW_MAJOR,
            use_height_and_width_as_shard_shape=True,
        )

        stats_mem_cfg = ttnn.create_sharded_memory_config(
            shape=(32, 32),
            core_grid=ttnn.CoreRangeSet([
                ttnn.CoreRange(ttnn.CoreCoord(0, 0), ttnn.CoreCoord(0, 0))
            ]),
            strategy=ttnn.ShardStrategy.WIDTH,
            orientation=ttnn.ShardOrientation.ROW_MAJOR,
            use_height_and_width_as_shard_shape=True,
        )

        replicate = ttnn.ReplicateTensorToMesh(mesh_device)

        input_torch = torch.randn(1, 1, 32, 4096, dtype=torch.bfloat16)
        input_tensor = ttnn.from_torch(
            input_torch,
            dtype=ttnn.bfloat16,
            layout=ttnn.TILE_LAYOUT,
            device=mesh_device,
            memory_config=input_mem_cfg,
            mesh_mapper=replicate,
        )

        gamma_torch = torch.randn(128, 32, dtype=torch.bfloat16)
        weight = ttnn.from_torch(
            gamma_torch,
            dtype=ttnn.bfloat16,
            layout=ttnn.ROW_MAJOR_LAYOUT,
            device=mesh_device,
            memory_config=ttnn.DRAM_MEMORY_CONFIG,
            mesh_mapper=replicate,
        )

        stats_torch = torch.zeros(1, 1, 32, 32, dtype=torch.float32)
        stats = ttnn.from_torch(
            stats_torch,
            dtype=ttnn.float32,
            layout=ttnn.TILE_LAYOUT,
            device=mesh_device,
            memory_config=stats_mem_cfg,
            mesh_mapper=replicate,
        )

        program_config = ttnn.LayerNormShardedMultiCoreProgramConfig(
            compute_with_storage_grid_size=(11, 6),
            subblock_w=1, block_h=1, block_w=2, inplace=False,
        )

        compute_kernel_config = ttnn.WormholeComputeKernelConfig(
            math_fidelity=ttnn.MathFidelity.HiFi4,
            math_approx_mode=False,
            fp32_dest_acc_en=True,
            packer_l1_acc=True,
        )

        global_sem = ttnn.create_global_semaphore(mesh_device, input_core_grid, 0)

        print(">>> Calling fused_rms_minimal — expect hang on Blackhole.", flush=True)
        out = ttnn.fused_rms_minimal(
            input_tensor,
            program_config,
            0,
            mesh_device,
            global_sem,
            topology=ttnn.Topology.Ring,
            memory_config=input_mem_cfg,
            epsilon=1e-5,
            weight=weight,
            stats=stats,
        )
        ttnn.synchronize_device(mesh_device)
        print(">>> Did NOT hang — got output shape:", out.shape)
    finally:
        ttnn.close_mesh_device(mesh_device)
        ttnn.set_fabric_config(ttnn.FabricConfig.DISABLED)


if __name__ == "__main__":
    main()
```

Verified independently in a clean tt-metal checkout — no progress past the "Calling fused_rms_minimal" print after several minutes.

**Comment by rpavlovicTT** [comment]
Hi @jonathansuTT , can you please take a look at this hang? If something is not configured properly, we should have a validation in the op so Forge compiler can discard such config.

**Comment by s-jovic** [comment]
I'll take a look since I had some recent fixes on this op after reduce helper changes

**Comment by rpavlovicTT** [comment]
PR #45152 fixes this. The root cause is the non-rectangular shard grid (64 workers in an 11×6=66-cell bounding box): `num_mcast_dests` was passed as `num_blocks=64` instead of `num_cores_x * num_cores_y=66`, so the NoC ack counter was undercredited and `noc_async_write_barrier()` waited forever. Verified locally — hang is gone with that patch applied.

**Comment by sshonTT** [comment]
https://github.com/tenstorrent/tt-metal/pull/45152 merged. I will close this issue.
